File: baekjoon/silver/1260_DFS와BFS/1260_DFS와BFS.py
#18
from collections import deque
def dfs(N,M,cur,visited,graph):
    visited.append(cur)
    print(cur,end=" ")
    for next in graph[cur]:
        if next not in visited:
            dfs(N,M,next,visited,graph)
    
    return
        



def bfs(N,M,cur,visited,graph):
    q = deque()
    q.append(cur)
    visited.append(cur)

    while q:
        c = q.popleft()
        print(c,end=" ")

        for next in graph[c]:
            if next not in visited:
                visited.append(next)
                q.append(next)


N, M, V = map(int,input().split())

dfs_visited = []
bfs_visited = []
dfs_graph = [[]for _ in range(N+1)]
bfs_graph = [[]for _ in range(N+1)]

for i in range(M):
    v1,v2 = map(int, input().split())
    dfs_graph[v1].append(v2)
    dfs_graph[v2].append(v1)
    bfs_graph[v1].append(v2)
    bfs_graph[v2].append(v1)

for i in range(N+1):
    dfs_graph[i] = sorted(dfs_graph[i])
    bfs_graph[i] = sorted(bfs_graph[i])
    

dfs(N,M,V,dfs_visited,dfs_graph)
print("")
bfs(N,M,V,bfs_visited,bfs_graph)
print("")

File: baekjoon/silver/1260_DFS와BFS/1260_DFS와BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;;

void DFS(int cur, vector<vector<int>> graph, vector<bool> &visited)
{
    visited[cur] = true;
    cout << cur << " ";

    for (int next : graph[cur])
    {
        if(!visited[next])
        {
            DFS(next, graph, visited);
        }
    }
}

void BFS(int start, vector<vector<int>> graph, vector<bool> visited)
{
    queue<int> q;

    q.push(start);
    visited[start] = true;
    cout << start << " ";

    while(!q.empty())
    {
        int cur = q.front();
        q.pop();

        for(int next : graph[cur])
        {
            if(!visited[next])
                {
                    cout << next << " ";
                    visited[next] = true;
                    q.push(next);
                }
        }
    }
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int N,M,V = 0;
    cin >> N >> M >> V;
    vector<vector<int>> graph(N+1);
    vector<bool> visited_D(N+1,false);
    vector<bool> visited_B(N+1,false);


    for (int i = 1; i <= M; i++)
    {
        int v1,v2;
        cin >> v1 >> v2;
        graph[v1].push_back(v2);
        graph[v2].push_back(v1);
    }

    for(int i = 1; i <= N; i++)
        sort(graph[i].begin(), graph[i].end());



    
    DFS(V, graph, visited_D);
    cout << endl;
    BFS(V, graph, visited_B);
    cout << endl;
    

    

    return 0;
}